### PR TITLE
2019-06-30 add v grammar.

### PR DESCRIPTION
**Pull request by ShenQingchuan**
# My V Grammar ANTLR4 RFC
Since I did some basic work for the V language and turns it into a `V.g4` file，
but it is not very complete, so I'm sending this pull request for comments.

### CODE REVIEW

**Comment by KvanTTT** [comment]
fixes #1464

File: v/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# V Language grammar for ANTLR4
+Because of the grammar similarity between Golang and V,
```
Comment (KvanTTT) [inline v/README.md]: Please always add a line break after headers (it's a common markdown style)

File: v/V.g4 [hunk 2]
```diff
@@ -0,0 +1,1332 @@
...
+        if (next.getChannel() == Token.DEFAULT_CHANNEL) {
+            // Keep track of the last token on the default channel.
+            this.lastToken = next;
+        }
+
+        return next;
+    }
+}
+
+//SourceFile       = ModuleClause ";" { ImportDecl ";" } { TopLevelDecl ";" } .
```
Comment (KvanTTT) [inline v/V.g4]: Maybe remove such comments? They are excess because parse rules are self-descriptive. Here and below.

File: v/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+It's modified from `golang.g4`
+
+Examples are all from [V language's official website](https://vlang.io)
+
+## Known Issues:
+- Because of the V is under early stage, its grammar is not so stable for now.
+
```
Comment (KvanTTT) [inline v/README.md]: Remove excess line break.

File: v/V.g4 [hunk 4]
```diff
@@ -0,0 +1,1332 @@
...
+sourceFile
+    : (moduleClause eos)? ( importDecl eos )* ( topLevelDecl eos )*
+    ;
+
+/*
+    V语言没有package关键字，取而代之的是 module 关键字
```
Comment (KvanTTT) [inline v/V.g4]: I don't like using of not English language in comments. We do international grammars 😃  Here and below.

File: v/V.g4 [hunk 5]
```diff
@@ -0,0 +1,1332 @@
...
+//ExpressionList = Expression { "," Expression } .
+expressionList
+    : expression ( ',' expression )*
+    ;
+
+
+
+
+
+
```
Comment (KvanTTT) [inline v/V.g4]: Excess line breaks. Here and below.

File: v/V.g4 [hunk 6]
```diff
@@ -0,0 +1,1332 @@
...
+    : declaration
+    | simpleStmt
+    | returnStmt
+    | breakStmt
+    | continueStmt
+    | block
+    | ifStmt
+    | switchStmt
+    | forStmt
+//    | labeledStmt     //TODO: V语言没有goto故也没有对应的label实现
```
Comment (KvanTTT) [inline v/V.g4]: Maybe just remove such not working code?

Comment (lijiansong) [reply v/V.g4]: Forgive my sharp words, such a terrible solution can be merged into master branch?

Now I wonder, does the other grammar rules in ANRLR been formalized proof？

Comment (KvanTTT) [reply v/V.g4]: > Forgive my sharp words, such a terrible solution can be merged into master branch?

That was done not by me.

> Now I wonder, does the other grammar rules in ANRLR been formalized proof？

What do you mean by formalized proof?

Comment (lijiansong) [reply v/V.g4]: A BNF rule for a certain programming language usually represents the type system of the PL, 
we may have to prove these rules is bug free by those PL theory. At least the rule is no ambiguity.

Comment (KvanTTT) [reply v/V.g4]: We only have example files that generated parsers from grammars should parse. It's far from formal proof, but it's enough for most cases.

Comment (ShenQingchuan) [reply v/V.g4]: ## My Apologize
I'm a sophomore student and just a ROOKIE in open-source community, 

@KvanTTT Thanks for all your kind suggestions 

and I would fix these AWFUL problems when I can get free from my current work.

I'm so sorry for my obtrusive PR but I have already mentioned that this PR is a RFC...

Thanks a lot for all your( @teverett ) work of failed examples founding !!


## For those grammar in comments...
@KvanTTT  Some V language grammars' implements are still on developing in V language community, and those grammars must be recorded for these language's future changes...

For example: `go` would be supported sooner, it was mentioned in the V language official documentation...

Comment (teverett) [reply v/V.g4]: Happy to help and thanks for your contribution to grammars-v4.

**Pull request merged**